### PR TITLE
update history page

### LINKED ISSUE #3455: add `payments` to history pages
@ehmatthes at https://github.com/gratipay/gratipay.com/issues/3446#issuecomment-103163666:

> I made an account for introtopython.org last week, and it was one of the teams that should have been included in last week's payday. But my balance is low enough that I didn't get a payout.
>
How will I know if my accounts are set up correctly, ie if my team was included in last week's payday? I don't know how to view my team account, or if it's even possible to view a profile page for that team yet. introtopython is not included in my teams page. My individual history doesn't show anything after payday 152.

#### Notify: 

- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2233
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2358
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2405
- [ ]  https://gratipay.freshdesk.com/helpdesk/tickets/2514
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2468
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2354
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2540

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/15744689-add-payments-to-history-pages?utm_campaign=plugin&utm_content=tracker%2F85909&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F85909&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by whit537**
Fixes #3455.

### CODE REVIEW

**Comment by whit537** [comment]
I'm thinking about this.

**Comment by whit537** [comment]
http://www.ledger-cli.org/

**Comment by whit537** [comment]
I'm thinking that Gratipay should be more normal accounting-wise. Double entry.

**Comment by whit537** [comment]
I want to avoid a rabbit hole, but can we move in that direction?

**Comment by whit537** [comment]
Properly speaking, every ~user, team, and payout route is an account.

**Comment by whit537** [comment]
We should probably have an `accounts` table and a `ledger` table. Makes me think of:

http://blog.balancedpayments.com/the-ledger/
http://blog.balancedpayments.com/state-machines/

**Comment by whit537** [comment]
Though I believe `journal` would be a better name than `ledger`:

http://www.accountingcoach.com/blog/general-ledger-general-journal.

**Comment by whit537** [comment]
https://en.wikipedia.org/wiki/Ledger
https://en.wikipedia.org/wiki/General_ledger
https://en.wikipedia.org/wiki/General_journal
https://en.wikipedia.org/wiki/Chart_of_accounts

**Comment by whit537** [comment]
> The ledger is a permanent summary of all amounts entered in supporting journals which list individual transactions by date. Every transaction flows from a journal to one or more ledgers.

https://en.wikipedia.org/wiki/Ledger

The journal is where double entry bookkeeping entries are recorded by debiting one or more accounts and crediting another one or more accounts with the same total amount. [...] In accounting and bookkeeping, a journal is a record of financial transactions in order by date.

https://en.wikipedia.org/wiki/General_journal

**Comment by whit537** [comment]
> A journal is also named the *book of original entry*, from when transactions were written in a journal prior to manually posting them to the accounts in the general ledger or subsidiary ledger.

https://en.wikipedia.org/wiki/General_journal

**Comment by whit537** [comment]
Let's say we make an `accounts` table. An account can be owned by more than one type of entity. How do we model that?

http://stackoverflow.com/questions/7844460/foreign-key-to-multiple-tables

**Comment by whit537** [comment]
http://www.accountingtools.com/debits-and-credits

**Comment by whit537** [comment]
> Accounts such as Cash, Investment Securities, and Loans Receivable are reported as assets on the bank's balance sheet. Deposits are reported as liabilities and include the balances in its customers' checking and savings accounts as well as certificates of deposit. In effect, your bank statement is just one of thousands of subsidiary records that account for millions of dollars in Deposits that a bank owes to its customers.

http://www.accountingcoach.com/debits-and-credits/explanation/4

**Comment by whit537** [comment]
"Basics of Escrow Accounting"
http://www.tdi.texas.gov/title/documents/BASICS.pdf

**Comment by whit537** [comment]
That's talking about regulated real estate escrow businesses. For our purposes the most relevant information is:

>**Escrow Accounting Formula**
>
Assets = Liabilities. There is no equity as in the normal accounting formula.

**Comment by whit537** [comment]
I am thinking about re-modeling Gratipay using proper double-entry accounting. We would fold both `exchanges` and `payments` (as well as the old `transfers` table) into a single `journal` table.

Gratipay's escrow would be fundamentally represented by an `escrow` asset account, offset by multiple `~user` liability accounts. Accounts would be defined relative to Gratipay:

| association             | account                  | type    | notes |
|:--------------------------|:-------------------------|:---------|:--|
|                                | escrow                   | asset   |   |
|                                | escrow receivable  | asset   | enables asynchronous payin routes |
|                                | escrow payable     | liability | enables asynchronous payout routes |
|                                | processing fee revenues | income | payments from our users |
|                                | processing fee expenses | expense | payments to our vendors (Braintree, PayPal) |
|                                | chargeback expenses | expense |  |
|                                | earned interest      | income |   |
| `participants`          | ~user                     | liability |   |
| `teams`                   | Team                     | liability |  |
| `exchange_routes` | payin route            | asset   | credit card |
| `exchange_routes` | payout route          | liability | PayPal |

We would implement Gratipay's business rules as triggers acting on the `journal`. So, for example, charging someone's credit card $10.00 would be a single `INSERT` of $10 into the `journal` for an account associated with a payin route (which in turn is associated with a ~user), and would result in the following `journal` entries:

| id (~=relative timestamp) | account                            |  debit | credit | 
|:---------------------------------:|-----------------------------------|--------:|--------:|
| 1                                       | cc-0                                  | 10.00 |    null |
| 2                                       | escrow receivable            |    null | 10.00 |
| 3                                       | escrow receivable            | 10.00 |    null |
| 4                                       | escrow                             |    null | 10.00 |
| 5                                       | escrow                             | 10.00 |    null |
| 6                                       | ~user                               |     null | 10.00 |
| 7                                       | ~user                               |    0.59 |    null |
| 8                                       | processing fee revenues |     null |   0.59 |

Though I'm not sure why we can't represent that like this:

| id (~=relative timestamp) | amount |  debit                     | credit                                | 
|:---------------------------------:|------------|-------------------------:|----------------------------------:|
| 1                                       | 10.00    | cc-0                       | escrow receivable            |
| 2                                       | 10.00    | escrow-receivable | escrow                             |
| 3                                       | 10.00    | escrow                   | ~user                               |
| 4                                       | 0.59      | ~user                     | processing fee revenues |

**Comment by whit537** [comment]
I'm staring at this, btw:

https://en.wikipedia.org/wiki/Debits_and_credits#Aspects_of_transactions

Also leaning heavily on:

http://www.accountingcoach.com/debits-and-credits/explanation/4

**Comment by whit537** [comment]
How about this for a fuller schema?

| id | amount |  debit                     | credit                               | parent | route | payday | recorder |
|:--:|-----------:|:-------------------------|:---------------------------------|:---------:|:-------:|:----------:|:-----------:|
| 1 | 10.00    | cash                       | ~user                               |   null    |     1    |      1     |     null    |
| 2 | 0.59      | ~user                     | processing fee revenues |      1     |   null  |      1     |     null    |

Let's have notes in a separate `journal_notes` table. That way we can have multiple notes on one journal entry, with some of the notes visible to users and others visible only to admins.

**Comment by whit537** [comment]
How would we handle exchange status with this schema?

**Comment by whit537** [comment]
Exchange statuses would be computable from the `journal`. Here's what a pending $10.00 payout would look like:

| id | amount |  debit                     | credit                               | parent |
|:--:|-----------:|:-------------------------|:---------------------------------|:---------:|
| 3 | 10.00    | ~user                     | escrow payable               |   null    |

Here's what a successful payout would look like:

| id | amount |  debit                     | credit                               | parent |
|:--:|-----------:|:-------------------------|:---------------------------------|:---------:|
| 3 | 10.00    | ~user                     | escrow payable               |   null    |
| 4 | 10.00    | escrow payable     | cash                                |      3     |

Here's what a failed payout would look like:

| id | amount |  debit                     | credit                               | parent |
|:--:|-----------:|:-------------------------|:---------------------------------|:---------:|
| 3 | 10.00    | ~user                     | escrow payable               |   null    |
| 4 | 10.00    | escrow payable     | ~user                               |      3     |

In other words, we'd need to query for child entries in order to compute transaction state. We could cache this computation in an additional `journal.status` field, and/or we could include a `journal.children` array in the schema.

**Comment by whit537** [comment]
Moving over to #3618.

**Comment by whit537** [comment]
Ready for review @rohitpaulk @rorepo @kzisme et al.

**Comment by whit537** [comment]
Alright, forging ahead with self-merge ...
